crates/uv/tests/it/run.rs

```diff
@@ -5500,3 +5500,49 @@ fn run_no_sync_incompatible_python() -> Result<()> {
 
     Ok(())
 }
+
+#[test]
+fn run_python_preference_no_project() {
+    let context =
+        TestContext::new_with_versions(&["3.12", "3.11"]).with_versions_as_managed(&["3.12"]);
+
+    context.venv().assert().success();
+
+    uv_snapshot!(context.filters(), context.run().arg("python").arg("--version"), @r"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    Python 3.12.[X]
+
+    ----- stderr -----
+    ");
+
+    uv_snapshot!(context.filters(), context.run().arg("--managed-python").arg("python").arg("--version"), @r"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    Python 3.12.[X]
+
+    ----- stderr -----
+    ");
+
+    // `VIRTUAL_ENV` is set here, so we'll ignore the flag
```

Comment:
How is `VIRTUAL_ENV` different here, doesn't that imply that I get different behavior depending on whether I activate the venv first and then `uv run` or whether I only use `uv run`, which is usually the same?

Reply:
`VIRTUAL_ENV` is set to the default virtual environment path (just by the test context)

This does imply you get different behavior based on whether the environment is active. I think that's correct though? With an active environment, it would be incorrect to ignore it (at most, we should warn), but when performing environment discovery, it makes sense to bypass environments that do not meet the preference.

Reply:
If you have a project with `requires-python = ">=3.11"` and try a too old venv with

```
uv venv -p 3.9
uv run python --version
```

then uv will recreate the venv in a compatible version and you'll get e.g. `Python 3.13.5`. This is independent of whether the venv is activated or not. I'd be surprised if there was different behavior between the Python version and the Python source wrt to the when we invalidate.

More concretely, I'd expect both blocks to share the invalidation behavior:

```
uv run --python-preference only-system -p 3.10 python --version
uv run --python-preference only-system -p 3.11 python --version
```

```
uv run --python-preference only-system -p 3.10 python --version
uv run --python-preference managed -p 3.10 python --version
```

Reply:
Yeah this behavior differs outside of projects because:

1. Outside of a project, we can't "invalidate" and recreate a virtual environment.
2. Projects don't consider `VIRTUAL_ENV` at all by default

This is intended to avoid making the breaking change too disruptive. While I might agree with you that it's better to have a consistent behavior, I think it's too risky to change.

Reply:
Oh, I see!